File: weekly-contest/192/4_1.py
class Solution
{
  int f[105][105][25];

public:
  int minCost (vector<int> &houses, vector<vector<int> > &cost, int m, int n,
	       int target)
  {
    int i, j, k, l, ans = 1 << 30;
    memset (f, 127, sizeof (f));
    f[0][0][0] = 0;
    for (i = 0; i < m; i++)
      for (j = 0; j <= i; j++)
	for (k = 0; k <= n; k++)
	  if (houses[i])
	    f[i + 1][j + (k != houses[i])][houses[i]]
	      = min (f[i + 1][j + (k != houses[i])][houses[i]], f[i][j][k]);
	  else
	    for (l = 1; l <= n; l++)
	      f[i + 1][j + (k != l)][l]
		= min (f[i + 1][j + (k != l)][l], f[i][j][k] + cost[i][l - 1]);
    for (l = 1; l <= n; l++)
      ans = min (ans, f[m][target][l]);
    if (ans == 1 << 30)
      ans = -1;
    return ans;
  }
};

File: weekly-contest/192/4_1.cpp
const int INF = 1000000000;
int f[111][22][111];

class Solution {
public:
    int minCost(vector<int>& houses, vector<vector<int>>& cost, int m, int n, int target) {
        for (int i = 0; i <= m; ++ i)
            for (int j = 0; j <= n; ++ j)
                for (int k = 0; k <= target; ++ k)
                    f[i][j][k] = INF;
        f[0][0][0] = 0;
        for (int i = 0; i < m; ++ i)
            for (int j = 0; j <= n; ++ j)
                for (int k = 0; k <= target; ++ k)
                {
                    if (f[i][j][k] == INF) continue;
                    if (houses[i] != 0)
                    {
                        int x = houses[i];
                        f[i+1][x][k+(j!=x)] = min(f[i+1][x][k+(j!=x)], f[i][j][k]);
                    }
                    else
                    {
                        for (int x = 1; x <= n; ++ x)
                        {
                            f[i+1][x][k+(j!=x)] = min(f[i+1][x][k+(j!=x)], f[i][j][k]+cost[i][x-1]);
                        }
                    }
                }
        int ret = INF;
        for (int j = 1; j <= n; ++ j)
            ret = min(ret, f[m][j][target]);
        if (ret == INF) ret = -1;
        return ret;
    }
};